### DISCUSSION TITLE
How does RabbitMQ 4 handle Webstomp Queues

### DISCUSSION THREAD

**Discussion by wetzesa** [rabbitmq/rabbitmq-server#12345] [Other] [answered]
### Community Support Policy

- [X] I have read [RabbitMQ's Community Support Policy](https://github.com/rabbitmq/rabbitmq-server/blob/main/COMMUNITY_SUPPORT.md)

### RabbitMQ version used

4.0.0-rc.1

### How is RabbitMQ deployed?

RPM package

### Steps to reproduce the behavior in question

Hi guys,

because of the release of RabbitMQ 4 and the deprecation notes I have an open Question about Webstomp.

As I know Webstomp Queues in RabbitMQ 3.x are defined as exclusive queues.
Now in RabbitMQ 4 exclusive queues are deprecated and the big question is, how it works now?

This means not the detailed implementation inside of RabbitMQ rather the concept and is there any relation which has to follow to use Webstomps in the new version or is it only an implementation change inside of the core of RabbitMQ 4 and we don't have to care about it?
Maybe there is a new library or property necessary or whatever ;-)

I could't find anything about this topic so I hope this thread will clarify the topic for all guys who use Webstomps with RabbitMQ.

Thanks!

**Comment by mkuratczyk** [comment, marked as answer]
Exclusive queues are not deprecated. Classic queues are not deprecated either. Mirroring is removed in 4.0 (after being deprecated for a few years).

Classic queues are better than ever, they are just not replicated: https://www.rabbitmq.com/blog/2024/01/11/3.13-release

**Reply by mkuratczyk** [reply]
exclusive queues were never mirrored by the way - since an exclusive queue is strictly coupled with the connection that created that queue, it is deleted when the connection terminates (if it wasn't, it'd just sit there, inaccessible by any connection). Therefore,  it makes no sense to maintain a mirror/replica of an exclusive queue, because you can't reconnect to another node and expect to see that queue.

**Comment by michaelklishin** [comment]
From `4.0.1` release notes:

```
Mirroring (replication) of classic queues https://github.com/rabbitmq/rabbitmq-server/pull/9815 after several years of deprecation. For replicated messaging data types,
use quorum queues and/or streams. Non-replicated classic queues remain and their development continues
```

"Non-replicated classic queues remain and their development continues" sounds pretty 

Exclusive queues are not mentioned once in the release notes. @wetzesa where are you getting this from, that "exclusive queues are deprecated"?

**Reply by wetzesa** [reply]
In the past there was a page with a table which show's the difference between classic queues and quorum queues.
It looks like, the page isn't any longer available because I can't find it any more. Maybe the page was revised and because of that I can't find it.

But anyways:
In the table on the page which I mean, there was in my memory spoken about classic and quorum queues. Not explict from mirrored classic queues.
Because of this memory and the table on the page were the exclusive feature marked as not available in quorum queues I were the conviction that the exclusive feature will be lost.
Additionally in the Web-UI all Webstomp queues are shown as classic queues with the additional flag "excl".
So it was not complete clear for me how Webstomps from RabbitMQ 4  working because of the table which I would like to show you but I can't found it.

For me it is now clear that my conviction based on the page was wrong and only the mirroring / replecation feature for classic queues is from RabbitMQ 4.x not longer available.

Thanks for the fast answer!

**Reply by mkuratczyk** [reply]
The page is here (it probably should be a separate page):
https://www.rabbitmq.com/docs/quorum-queues#feature-matrix

**Reply by wetzesa** [reply]
Ah got it...
I seen this matrix round about 6 to 12 months ago and for me it looked like that the exclusive feature is bounded to the classic queues. That was the only reason why I got trouble with this topic but now it is clear

**Reply by mkuratczyk** [reply]
The "exclusive" feature is specific to classic queues - there is no point in having an exclusive quorum queue, so it's not allowed.
An exclusive quorum queue, if it existed, would simply be a slower exclusive queue with no benefits over an exclusive classic queue.
